packages/dd-trace/src/appsec/iast/taint-tracking/operations-taint-object.js

```diff
@@ -40,6 +40,55 @@ function taintObject (iastContext, object, type) {
   return result
 }
 
+function taintQueryWithCache (iastContext, query, type) {
+  const transactionId = iastContext?.[IAST_TRANSACTION_ID]
+  if (!transactionId || !query) return
+
+  if (!iastContext.queryCache) {
+    iastContext.queryCache = {}
+  }
+
+  return traverseAndTaint(query, iastContext.queryCache, '', type, transactionId)
+}
+
+function traverseAndTaint (value, cache, path, type, transactionId) {
+  if (value == null) return value
+
+  if (typeof value === 'string') {
+    // If we already have a tainted version of this exact string in the cache
+    if (value === cache) {
+      return cache
+    }
+
+    return TaintedUtils.newTaintedString(transactionId, value, path, type)
+  }
+
+  if (typeof value === 'object') {
+    const isArray = Array.isArray(value)
+    if (!cache || typeof cache !== 'object' || Array.isArray(cache) !== isArray) {
+      cache = isArray ? [] : Object.create(null)
+    }
+
+    for (const key of Object.keys(value)) {
+      const childPath = path ? `${path}.${key}` : key
+      const childValue = value[key]
+
+      value[key] = traverseAndTaint(
+        childValue,
+        cache[key],
+        childPath,
+        type,
+        transactionId
+      )
+
+      cache[key] = value[key]
```

Comment:
If an application modifies `req.query` object, the cache will be also updated:

```
(req, res) => {
  req.query.paramThatNotExists = 'AAAA'
// or also
  req.query.paramThatExists = 'aaaa' // => our cached object will be updated
}
```

our object in cache must be different instances of objects that are in the original query object, only the string instances can be the same instances.